caddyhttp/proxy/reverseproxy.go

```diff
@@ -126,6 +127,35 @@ func NewSingleHostReverseProxy(target *url.URL, without string) *ReverseProxy {
 	return rp
 }
 
+func NewDynamicHostReverseProxy(templateUrl string, without string) *ReverseProxy {
+	director := func(req *http.Request) {
+		replacer := httpserver.NewReplacer(req, nil, "")
+
+		templateUrl = replacer.Replace(templateUrl)
+
+		target, _ := url.Parse(templateUrl)
+		targetQuery := target.RawQuery
+
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = replacer.Replace(strings.Replace(target.Host, "HOST", "{host}", -1))
```

Comment:
How about doing the replacement before parsing, then you won't have to hack around the { } like this?